### PR TITLE
DOC: How to verify your first NumPy bug

### PR DESCRIPTION
**Pull request by drskoolie**
In the last community meeting, @rgommers suggested to start with this how-to first. 

Motivation: Beginners who are looking to contribute to NumPy don't know where to start. This document is the first in a series to ease them in gently. 

Summary of how-to:
1. Setting up a virtual environment
2. Installing the appropriate version of NumPy
3. Verifying the existence of the bug
4. Uninstalling NumPy and installing the latest version + Confirm the bug has been solved

Issues:
1. Don't know how to cross reference links in Sphinx, **line 61** needs to be replaced with the appropriate link

### CODE REVIEW

File: doc/source/user/how-to-verify-bug.rst [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    pip install numpy==1.18.4
+
+******************************************************************************
+3. Confirm their bug exists
+******************************************************************************
+
+So their claim is simple enough, they say that the wrong dtype is returned
+if one of the inputs of the method `np.polymul` is a zero array. Open up a
+python terminal and type their code in and see if you can reproduce the
+bug:
```
Comment (melissawm) [inline doc/source/user/how-to-verify-bug.rst:76]:
```suggestion
Their claim is simple enough: the wrong dtype is returned 
if one of the inputs of the method `numpy.polymul` is a zero array. Open up a
Python terminal and type their code in and see if you can reproduce the
bug:
```

File: doc/source/user/how-to-verify-bug.rst [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+Although for this specific bug, it has already been solved here_. To confirm,
+first uninstall the old version of NumPy and install the latest one:
+
+.. _here: https://github.com/numpy/numpy/pull/17577
+
+::
+
+    pip uninstall numpy
+    pip install numpy
+
```
Comment (melissawm) [inline doc/source/user/how-to-verify-bug.rst:109]: Might be worth mentioning that if you're using venv or conda you can actually have parallel installations with different versions of NumPy.

File: doc/source/user/how-to-verify-bug.rst [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    dtype('complex64')
+
+******************************************************************************
+5. What to do next
+******************************************************************************
+
+Go to the NumPy issues page https://github.com/numpy/numpy/issues and see if
+you can confirm the existence of any other bugs. If you can, comment on the 
+issue saying so. Doing this helps the NumPy developers know if more than one
+user is experiencing the issue.
```
Comment (melissawm) [inline doc/source/user/how-to-verify-bug.rst:136]:
```suggestion
Go to the `NumPy GitHub issues page <https://github.com/numpy/numpy/issues>`_
and see if you can confirm the existence of any other bugs. If you can, comment
on the issue saying so. Doing this helps the NumPy developers know if more than
one user is experiencing the issue.
```

File: doc/source/user/how-to-verify-bug.rst [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+4. See if the bug still exists on the latest version of NumPy
+******************************************************************************
+
+Now that the bug is confirmed, you could try and solve it by opening a PR.
+Although for this specific bug, it has already been solved here_. To confirm,
+first uninstall the old version of NumPy and install the latest one:
+
+.. _here: https://github.com/numpy/numpy/pull/17577
+
+::
```
Comment (melissawm) [inline doc/source/user/how-to-verify-bug.rst:105]:
```suggestion
Now that the bug is confirmed, you could try and solve it by opening a PR.
Although for this specific bug, it has already been solved 
`here <https://github.com/numpy/numpy/pull/17577>`_. To confirm,
first uninstall the old version of NumPy and install the latest one::
```

(note that using two colons on the last line before the code block is enough and you don't have to add the double colon on a separate line)

File: doc/source/user/how-to-verify-bug.rst [hunk 5]
```diff
@@ -0,0 +1,136 @@
+.. _how-to-verify-bug:
+
+##############################################################################
+How to verify your first NumPy bug
+##############################################################################
+
+This how-to will go through the process of verifying a bug_ that was submitted
+on the Github issues page.
+
+.. _bug: https://github.com/numpy/numpy/issues/16354
```
Comment (melissawm) [inline doc/source/user/how-to-verify-bug.rst:10]:
```suggestion
This how-to will go through the process of verifying a
`bug <https://github.com/numpy/numpy/issues/16354>`_ that was submitted
on the Github issues page.
```

**Review by melissawm — CHANGES_REQUESTED** [review]
Thanks, @a-elhag ! This looks like a great starter, I left some comments mostly about the sphinx directives. Let me know if you have questions!

**Comment by a-elhag** [comment]
Thanks @melissawm for the review! I've tried to address the points and the only one that is left is about **Section 2: Install the NumPy version mentioned in the bug report**. Do let me know if there are anymore changes that need to be made.

**Comment by melissawm** [comment]
Hi @a-elhag ! There's some work to be done here - the CI failures are coming from the doctests on your document. Can you take a look? Thanks :)

**Comment by seberg** [comment]
It appears that this PR was incorrectly closed by github while renaming the NumPy main (previously master) branch. I suspect because the original repository was also deleted.  That seems like a github issue, but I am not sure there is a solution to it aside from opening a new PR.

Sorry about that issue, I did not review the PR. However, even if the local branches were deleted, it seems the work is still available (e.g. using `https://github.com/numpy/numpy/pull/<pull-request-number>` with the command line interface), so we can assist creating a new PR!

**Comment by InessaPawson** [comment]
@a-elhag, thank you for working on this PR! It would be awesome if you could finish it. It’s such an important topic! As Sebastian said, we can help you to restore the work you submitted.

**Comment by rgommers** [comment]
I suggest taking over this PR @InessaPawson and opening a new one to get this valuable content merged. Here's how I would do that:
1. Install https://cli.github.com/. This will give you the `gh` tool which helps dealing with PRs.
2. Assuming you have your repos set up so that `upstream` points to the main NumPy repo and `origin` to your own fork, do:

```bash
$ git checkout main
$ git pull  # ensure your main branch is up to date
$ gh pr checkout 17851  # this may ask you which repo to use the first time, choose numpy/numpy
$ git rebase -i HEAD^^^^

# Squash commits:
pick c103615e3 DOC: How to verify your first numpy bug
squash ae6f5b9e5 Apply suggestions from code review
squash adb0005e3 minor formatting changes
squash 225373a96 simple formating

# Use commit message:
DOC: How to verify your first numpy bug

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

$ git rebase main  # This will give one conflict, resolve it and then do:
$ git add doc/source/user/howtos_index.rst
$ git rebase --continue  # now you have a new branch with a single commit on top of main
$ # address remaining content issues from review on this PR, add those as a new commit
$ git push origin howto-bug  # push changes to your own fork; after this make a new PR
```

**Comment by PrecisionWordcraft** [comment]
@melissawm Shall I make review comments in this PR or should I fork and then follow the process outlined above by Ralf?

File: doc/source/user/how-to-verify-bug.rst [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    pip install numpy
+
+
+Now run the code again and confirm that the bug was solved.
+
+::
+
+    >>> import numpy as np
+    >>> np.__version__
+    '1.18.4'
```
Comment (PrecisionWordcraft) [inline doc/source/user/how-to-verify-bug.rst:111]: This version number is the same as the one above where the bug exists. That can't be the case.

I would resolve this by noting in the running text of the doc that the version of numpy will be the latest, and that's what you should expect to see here.

File: doc/source/user/how-to-verify-bug.rst [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+Their claim is simple enough: the wrong dtype is returned if one of the inputs
+of the method `numpy.polymul` is a zero array. Open up a Python terminal and
+type their code in and see if you can reproduce the bug::
+
+    >>> import numpy as np
+    >>> np.__version__
+    '1.18.4'
+    >>> a = np.array([1,2,3])
+    >>> z = np.array([0,0,0])
+    >>> np.polymul(a.astype(np.int64), a.astype(np.int64)).dtype
```
Comment (PrecisionWordcraft) [inline doc/source/user/how-to-verify-bug.rst:82]: Some information as to why the returned data types aren't correct might be helpful.

File: doc/source/user/how-to-verify-bug.rst [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+::
+
+    virtualenv venv_np_bug
+    source venv_np_bug/bin/activate
+
+******************************************************************************
+2. Install their version of NumPy
+******************************************************************************
+
+If the bug was more complex, you may need to build the NumPy version they
```
[the remark this answers is not in the document]
Comment (brandondavid) [reply doc/source/user/how-to-verify-bug.rst:63]: I would suggest _"If the bug is more complex, you may need to.."_ or _"For complex bugs, you may need to..."_

**Pull request closed without merging**